### DISCUSSION TITLE
How to install FFmpeg to stability matrix?

### DISCUSSION THREAD

**Discussion by ziilch** [LykosAI/StabilityMatrix#676] [Q&A]
I'm trying to do some cool stuff with animate diff combined with A detailer and control net combined with textual inversions and loras. 

Currently I'm encountering an issue where inside the stability matrix terminal I notice it's saying it can't find FFMPEG and its reverting to OpenCV in this line of code:
----------------------

AnimateDiff - ERROR - [AnimateDiff] Error extracting frames via ffmpeg: [WinError 2] The system cannot find the file specified, fall back to OpenCV.

-------------------------
Theres a ton of other code and errors like :
-------------------------

C:\actions-runner\_work\pytorch\pytorch\builder\windows\pytorch\aten\src\ATen\native\cuda\IndexKernel.cu:92: block: [814,0,0], thread: [56,0,0] Assertion `-sizes[i] <= index && index < sizes[i] && "index out of bounds"` 

0%|          | 0/20 [00:00<?, ?it/s]MemMon:
Traceback (most recent call last):
  File "threading.py", line 1016, in _bootstrap_inner

  File "D:\Stability Matrix\Data\Packages\Stable Diffusion WebUI\modules\memmon.py", line 53, in run
    free, total = self.cuda_mem_get_info()
  File "D:\Stability Matrix\Data\Packages\Stable Diffusion WebUI\modules\memmon.py", line 34, in cuda_mem_get_info
    return torch.cuda.mem_get_info(index)
  File "D:\Stability Matrix\Data\Packages\Stable Diffusion WebUI\venv\lib\site-packages\torch\cuda\memory.py", line 663, in mem_get_info
    return torch.cuda.cudart().cudaMemGetInfo(device)

------------------------
and then I get a cuda error:
-----------------------------

RuntimeError: CUDA error: device-side assert triggered
CUDA kernel errors might be asynchronously reported at some other API call, so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1.
Compile with `TORCH_USE_CUDA_DSA` to enable device-side assertions.

-------------------------------------

Now I am easily able to install FFMPEG within the version of python I have associated with the default file path when installing python into windows 11, I used a pip command and got it installed no problem using the various tutorials online. The issue is that doesn't interact with anything within stability diffusions web ui that stability matrix installed because its using a virtual version of python within the venv folder within the stable diffusion webui a1111 file structure. I also notice stability matrix is using python 3.10 and the version of python I personally installed onto windows is 3.12 so already it shows me they are entirely different installations of python.

Anytime I open up the python interpreter within the stability matrix file structure and try any of the get or pip commands it says I have a syntax error. I have a hunch I'm in the wrong terminal to perform those commands.

I've also tried going to settings within stability matrix and edit the environment variables to include the installation path for FFmpeg and that doesn't seem to work (though maybe I'm editing it incorrectly, you tell me)

I know I'm not the only one on the web with this issue, and honestly I'm confused why ffmpeg isn't included in stability matrix's installation of a1111 and its virtual version of python, seems like a critical thing a lot of people need and there isn't clear instructions on line on how to fix it specifically with stability matrix. For a system thats supposed to be easier to work with, this particular aspect has been pretty frustrating and you would think this would have been included at installation without needing extra effort from the end user.

Any help would be greatly appreciated and I'll pay it forward and relay the information on reddit and elsewhere on github to the other users who asked regarding this problem.

**Comment by drdancm** [comment]
On the subject of FFmpeg, some programs such as FaceFusion, require it, so I don't know if StabMatrix is or can be involved in installing this. Possibly, it depends on whether all packages may have access to this, or whether each package has to deal with it on its own.

FaceFusion also wants to have conda installed, so I have the same questions about that also.

**Comment by ziilch** [comment]
I ended up figuring it out, in order to install ffmpeg with stability matrix. You go to the ffmpeg website, 

https://www.ffmpeg.org/download.html

get the installation files, 

for me I have windows so its here:
https://www.gyan.dev/ffmpeg/builds/

select "ffmpeg-git-essentials.7z"

and once you download then unzip that package, put the unzipped files (no folders) directly into the root of stable diffusion web ui folder within your stability matrix installation location. 

For me its D:\Stability Matrix\Data\Packages\Stable Diffusion WebUI

once you put them in the root folder it will be able to see it. Probably will also want to restart stability matrix to let it fully start over and refresh.

**Comment by drdancm** [comment]
Thanks very much for your detailed instruction except for 

"no folders"

I assume you mean not the top folder (ffmpeg-2024-07-24-git-896c22ef00-essentials_build), but copy everything that is under it including the folders which have contents.

The following stuff which have contents are:
bin (folder)
doc (folder)
presets (folder)
LICENSE
README.txt

Is that correct?

**Reply by ziilch** [reply]
everything within, I believe the bin folder, but not the folder itself, just the contents

The particular files I have is:
ffmpeg.exe
ffplay.exe
ffprobe.exe

Once these are in the root of your stable diffusion web ui, as previously stated for me its:

 D:\Stability Matrix\Data\Packages\Stable Diffusion WebUI
 
 Once those 3 files are in that root folder your errors will be resolved
 
 Cheers

**Comment by LadyFlames** [comment]
easier solution for installing things for webui´s there on packages you have 3 dots at the top next to the name of the webui just press  that and then go python packages then press the Plus symbol in there just type the one you want to install its that simple and if you want to uninstall something press the Minus in there and if you want to upadate or downgrade versions just press the on the name of the thing you want to update or downgrade it will have a version number it you press that it will show if there is a new version available and it will show either up arrow if update is available or it will show down arrow if you want to downgrade the version
![Näyttökuva (61)](https://github.com/user-attachments/assets/0b9a5599-1f59-4a3c-933a-433c6c06d1fa)
![Näyttökuva (62)](https://github.com/user-attachments/assets/1d952097-17ac-4561-9531-9b3846b33759)
![Näyttökuva (63)](https://github.com/user-attachments/assets/10829076-6cee-44ab-8cd5-d3c349c81a69)
